src/App/Utilities/MapToolExporter.ts

```diff
@@ -0,0 +1,740 @@
+    attr3: (fromMaybe ("MU", List.subscript (Skill.A.check (skillFromWiki)) (2)) as MaptoolsAttribute),
+  })
+}
+
+function buildSpell (spellFromWiki: Record<Spell>):
+(spellFromHero: Maybe<Record<SkillDependent>>) => MaptoolsSkill {
+  return spellFromHero => ({
+    name: Spell.A.name (spellFromWiki),
+    value: isJust (spellFromHero) ? SkillDependent.A.value (spellFromHero.value) : 0,
+    attr1: (fromMaybe ("MU", List.subscript (Spell.A.check (spellFromWiki)) (0)) as MaptoolsAttribute),
+    attr2: (fromMaybe ("MU", List.subscript (Spell.A.check (spellFromWiki)) (1)) as MaptoolsAttribute),
+    attr3: (fromMaybe ("MU", List.subscript (Spell.A.check (spellFromWiki)) (2)) as MaptoolsAttribute),
+  })
+}
+
+function buildChant (chantFromWiki: Record<LiturgicalChant>):
+(chantFromHero: Maybe<Record<SkillDependent>>) => MaptoolsSkill {
+  return chantFromHero => ({
+    name: LiturgicalChant.A.name (chantFromWiki),
+    value: isJust (chantFromHero) ? SkillDependent.A.value (chantFromHero.value) : 0,
+    attr1: (fromMaybe ("MU", List.subscript (LiturgicalChant.A.check (chantFromWiki)) (0)) as MaptoolsAttribute),
+    attr2: (fromMaybe ("MU", List.subscript (LiturgicalChant.A.check (chantFromWiki)) (1)) as MaptoolsAttribute),
+    attr3: (fromMaybe ("MU", List.subscript (LiturgicalChant.A.check (chantFromWiki)) (2)) as MaptoolsAttribute),
+  })
+}
+
+function getTalentsXML (hero: HeroModelRecord, state: AppStateRecord): string {
+  const buildSkillList =
+  pipe (
+    SkillGroup.A.id,
+    grid =>
+      pipe_ (state.values.wiki.values.skills,
+        OrderedMap.filter (skill => Skill.A.gr (skill) === grid),
+        OrderedMap.elems,
+        fmap ((skill: Record<Skill>) => pipe_ (skill,
+          Skill.A.id,
+          OrderedMap.lookupF (hero.values.skills),
+          buildSkill (skill))))
+  )
+
+  const skills: List<MapToolsSkillGroup> =
+  pipe_ (state.values.wiki.values.skillGroups,
+    OrderedMap.elems,
+    fmap (
+      (g: Record<SkillGroup>) => ({
+        groupName: skillGroupNameForMapTool (g),
+        skills: buildSkillList (g),
+      })
+  ))
+
+  const xml = `[${
+    pipe_ (skills,
+      fmap ((sg: MapToolsSkillGroup) => ({
+        key: sg.groupName,
+        value: pipe_ (sg.skills,
+          fmap ((skill: MaptoolsSkill) =>
+            `{"Talent":"${skill.name}","Talentwert":${skill.value}},"Probe":{`
+            + `"Eigenschaft1":"${skill.attr1}",`
+            + `"Eigenschaft2":"${skill.attr2}",`
+            + `"Eigenschaft3":"${skill.attr3}"}}`)),
+    })))
+   }]`
+
+  // TODO: Zauber/Rituale und Liturgien/Zeremonien noch herausfinden
+  // AsP- bzw. KaP-Kosten könnte man auch exportieren
+
+  const spells =
+    pipe_ (state.values.wiki.values.spells,
+      OrderedMap.elems,
+      fmap ((s: Record<Spell>) => pipe_ (
+        s,
+        Spell.A.id, // TODO: Typfehler?
+        Maybe.mapMaybe (OrderedMap.lookupF (hero.values.spells)),
+        buildSpell (s)
+    )))
```

Comment:
Hier versuchst du, über einen String zu mappen. Stattdessen könntest du nur die IDs extrahieren und dann über die Liste mappen.

```suggestion
    pipe_ (
      state.values.wiki.values.spells,
      OrderedMap.elems,
      Maybe.mapMaybe ((s: Record<Spell>) => pipe_ (
        s,
        Spell.A.id,
        OrderedMap.lookupF (hero.values.spells),
        fmap (buildSpell (s))
      ))
    )
```

Ich würde aber vorschlagen, es umgekehrt zu machen: Über die Daten des Helden mappen und dann die dazugehörigen Einträge im Wiki suchen. Denn: Der Held hat meist nur einen Bruchteil der verfügbaren Optionen. Du iterierst hier also mehrfach durch sehr lange Listen. Daher, mein Vorschlag wäre:

```suggestion
    pipe_ (
      hero,
      Hero.A.spells,
      OrderedMap.elems,
      Maybe.mapMaybe (heroSpell => pipe_ (
        heroSpell,
        ActivatableSkillDependent.A.id,
        OrderedMap.lookupF (StaticData.A.spells (wiki)),
        fmap (staticSpell => buildSpell (staticSpell) (heroSpell))
      ))
    )
```

Bei Liturgien die Zeilen weiter unten müsste das Problem auf gleiche Weise gelöst werden können. 👍🏻

Reply:
Was ich noch wissen müsste ist wie ich die Zauber und Liturgien entsprechend gruppieren kann. Ich hätter Zauber und Rituale gerne getrennt. Gleiches bei Liturgien und Zeremonien.

Reply:
Die haben eine `gr`-Property, danach kannst du gruppieren. `1` sind Zaubersprüche/Liturgien und `2` sind Rituale/Zeremonien. Das Filtern müsstest du zum Beispiel mit `List.partition` realisieren können.

Reply:
Gibt es außer den beiden noch weitere Gruppen, die ich vielleicht beachten muss? Die Gruppierung erfolgt ja weil für die beiden Typen evtl. unterschiedliche Regelungen gelten, die ich in maptools beachten muss. So wird bei Ritualen z.B. die Probe um passende Kleidung modifiziert, was bei Zaubern nicht der Fall ist.

Reply:
Und wie bekomme ich nach dem partition die 1. und die 2. Subliste?

Reply:
> Gibt es außer den beiden noch weitere Gruppen, die ich vielleicht beachten muss? Die Gruppierung erfolgt ja weil für die beiden Typen evtl. unterschiedliche Regelungen gelten, die ich in maptools beachten muss. So wird bei Ritualen z.B. die Probe um passende Kleidung modifiziert, was bei Zaubern nicht der Fall ist.

Oh wait. In den `spells` des Helden sind **noch** die magischen Handlungen gespeichert. Dann geht das gar nicht über `List.partition`, sondern eher mit `List.groupByKey`. Da es dann ja mehr als zwei Gruppen sind. Welche ID welcher Gruppe zugeordnet ist, kannst du in der Datenbank auch den `SpellGroups` entnehmen.

Um deine zweite Frage trotzdem zu beantworten: Aus einem Paar kannst du mit `Tuple.fst` und `Tuple.snd` die Werte holen.

Reply:
List.partition scheint aber zu funktionieren. Ändert sich da in Zukunft was oder kann ich das so lassen?

Reply:
In Zukunft sind die Gruppen komplett getrennt voneinander. Dann brauchst du gar nicht mehr filtern. Bis dahin kannst du das so lassen! 👍🏻 Das werde ich dann wahrscheinlich auch selbst anpassen.

Reply:
Alles klar. Aktuell werden halt noch keine Zaubertricks und Segnungen exportiert. Ist jetzt kein Beinbruch wenn das noch nicht drin ist. Wäre aber nice.